scipy/signal/tests/test_signaltools.py

```diff
@@ -2172,99 +2406,138 @@ def _setup_rank1(self, dt, mode):
         y = correlate(a, b, 'valid')
-        assert_array_almost_equal(y, y_r, decimal=self.decimal(dt))
-        assert_equal(y.dtype, dt)
+        assert_array_almost_equal(y, y_r, decimal=self.decimal(dt, xp))
+        assert y.dtype == dt
 
         # See gh-5897
         y = correlate(b, a, 'valid')
-        assert_array_almost_equal(y, y_r[::-1].conj(), decimal=self.decimal(dt))
-        assert_equal(y.dtype, dt)
+        flip = array_namespace(y_r).flip
+        assert_array_almost_equal(y, xp.conj(flip(y_r)),
+                                  decimal=self.decimal(dt, xp))
+        assert y.dtype == dt
+
+    def test_rank1_same(self, dt_name, xp):
+        a, b, y_r = self._setup_rank1(dt_name, 'same', xp)
+        dt = getattr(xp, dt_name)
 
-    def test_rank1_same(self, dt):
-        a, b, y_r = self._setup_rank1(dt, 'same')
         y = correlate(a, b, 'same')
-        assert_array_almost_equal(y, y_r, decimal=self.decimal(dt))
-        assert_equal(y.dtype, dt)
+        assert_array_almost_equal(y, y_r, decimal=self.decimal(dt, xp))
+        assert y.dtype == dt
 
-    def test_rank1_full(self, dt):
-        a, b, y_r = self._setup_rank1(dt, 'full')
+    def test_rank1_full(self, dt_name, xp):
+        a, b, y_r = self._setup_rank1(dt_name, 'full', xp)
+        dt = getattr(xp, dt_name)
         y = correlate(a, b, 'full')
-        assert_array_almost_equal(y, y_r, decimal=self.decimal(dt))
-        assert_equal(y.dtype, dt)
+        assert_array_almost_equal(y, y_r, decimal=self.decimal(dt, xp))
+        assert y.dtype == dt
 
-    def test_swap_full(self, dt):
-        d = np.array([0.+0.j, 1.+1.j, 2.+2.j], dtype=dt)
-        k = np.array([1.+3.j, 2.+4.j, 3.+5.j, 4.+6.j], dtype=dt)
+    def test_swap_full(self, dt_name, xp):
+        dt = getattr(xp, dt_name)
+        d = xp.asarray([0.+0.j, 1.+1.j, 2.+2.j], dtype=dt)
+        k = xp.asarray([1.+3.j, 2.+4.j, 3.+5.j, 4.+6.j], dtype=dt)
         y = correlate(d, k)
-        assert_equal(y, [0.+0.j, 10.-2.j, 28.-6.j, 22.-6.j, 16.-6.j, 8.-4.j])
-
-    def test_swap_same(self, dt):
-        d = [0.+0.j, 1.+1.j, 2.+2.j]
-        k = [1.+3.j, 2.+4.j, 3.+5.j, 4.+6.j]
+        xp_assert_close(
+            y, xp.asarray([0.+0.j, 10.-2.j, 28.-6.j, 22.-6.j, 16.-6.j, 8.-4.j]),
+            atol=1e-6, check_dtype=False
+        )
+
+    def test_swap_same(self, dt_name, xp):
+        d = xp.asarray([0.+0.j, 1.+1.j, 2.+2.j])
+        k = xp.asarray([1.+3.j, 2.+4.j, 3.+5.j, 4.+6.j])
         y = correlate(d, k, mode="same")
-        assert_equal(y, [10.-2.j, 28.-6.j, 22.-6.j])
+        xp_assert_close(y, xp.asarray([10.-2.j, 28.-6.j, 22.-6.j]))
 
-    def test_rank3(self, dt):
-        a = np.random.randn(10, 8, 6).astype(dt)
-        a += 1j * np.random.randn(10, 8, 6).astype(dt)
-        b = np.random.randn(8, 6, 4).astype(dt)
-        b += 1j * np.random.randn(8, 6, 4).astype(dt)
+    @skip_xp_backends("cupy", reason="notimplementederror")
+    def test_rank3(self, dt_name, xp):
```

Comment:
Np only?

Reply:
Not necessatily, why. It working on torch and jax.numpy, just cupy raises a NotImplementedError for ndim > 1.